### PR TITLE
Create a single dependencies file and create cpanfile/OpenQA.spec

### PR DESCRIPTION
**Pull request by perlpunk**
Issue: https://progress.opensuse.org/issues/56525

### CODE REVIEW

**Comment by okurz** [comment]
Interesting. But just to be clear: we also have the dependencies.txt introduced by Andrii which motivated us for the task to specify deps at less places. In the state as is we now have one more place where deps are mentioned even though they can be auto-generated. We are looking for a way to have at least one place left :)

**Comment by perlpunk** [comment]
@okurz If my script would be able to also generate dependencies.txt, then it would be ok?

**Comment by okurz** [comment]
It can even be ok without it because it already helps. Maybe we can add a caution header to the files that are should be updated automatically with a hint how to do so. In the best case we never have duplicate entries committed. E.g. there is a way to shell shell context from the header of spec files

**Comment by perlpunk** [comment]
I added a comment/usage in `dependencies.yaml`, and autogenerated comments in `cpanfile` and above every generated line in `openQA.spec`.

**Comment by perlpunk** [comment]
> * Call the dependencies update also as part of tests and fail tests if there is a git diff on cpanfile or openQA.spec

yep, I thought about that too

File: Makefile [hunk 1]
```diff
@@ -252,3 +252,6 @@ test-yaml:
 	@which yamllint >/dev/null 2>&1 || echo "Command 'yamllint' not found, can not execute YAML syntax checks"
 	@# Fall back to find if there is no git, e.g. in package builds
 	yamllint --strict $$((git ls-files "*.yml" "*.yaml" 2>/dev/null || find -name '*.y*ml') | grep -v ^dbicdh)
+
+update-deps:
+	perl etc/update-deps.pl
```
Comment (okurz) [inline Makefile:257]: @perlpunk your change inspired me to do https://github.com/os-autoinst/openQA/pull/2939 . Hence I suggest to provide update-deps.pl not in "etc" which should be reserved for configuration files to be installed on systems but either – after my PR – in tools or in script and then deleted on install in Makefile

File: etc/update-deps.pl [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+            $specline .= " $version";
+        }
+    }
+    my $comment = "# The following line is generated from $yamlfile";
+    if ($spec =~ s/^# .*generated.*\n^$prefix.*/$comment\n$specline/m) {
+        next;
+    }
+    # No comment above the line yet
+    unless ($spec =~ s/^$prefix.*/$comment\n$specline/m) {
+        warn "/^$prefix/ not found in $specfile";
```
Comment (kalikiana) [inline etc/update-deps.pl:45]: This should probably fail? Considering the list of targets is explicit

Comment (perlpunk) [reply etc/update-deps.pl]: right. fixed!

File: Makefile (hunk 1 above)
Comment (perlpunk) [reply Makefile]: moved it to `tools/`.
Do I need to delete it explicitly or is this directory already removed on install?

Comment (kalikiana) [reply Makefile]: I see no delete happening in #2939/master

Comment (okurz) [reply Makefile]: Exactly, not needing an explicit mention what was I wanted to achieve

**Comment by perlpunk** [comment]
I added a test that runs `make update-deps` and checks the git status.

File: Makefile [hunk 3]
```diff
@@ -252,3 +252,6 @@ test-yaml:
 	@which yamllint >/dev/null 2>&1 || echo "Command 'yamllint' not found, can not execute YAML syntax checks"
 	@# Fall back to find if there is no git, e.g. in package builds
 	yamllint --strict $$((git ls-files "*.yml" "*.yaml" 2>/dev/null || find -name '*.y*ml') | grep -v ^dbicdh)
+
+update-deps:
```
[the remark this answers is not in the document]
Comment (perlpunk) [reply Makefile:252]: done

File: tools/update-deps [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+};
+EOM
+
+    $cpan .= <<'EOM';
+
+feature 'coverage', 'coverage for CI' => sub {
+    requires 'Devel::Cover';
+    requires 'Devel::Cover::Report::Codecov';
+};
+EOM
```
Comment (okurz) [inline tools/update-deps:116]: I suggest to put this block into a file so that we do not need to ask contributors to change this script in case we change something about dependencies.

Comment (perlpunk) [reply tools/update-deps]: ok, I added a `tools/cpanfile.additions`

**Comment by perlpunk** [comment]
I also had a look at `.circle/dependencies.txt`.
Currently it is generated by installing openQA (which includes os-autoinst), and then listing all installed packages. So it includes additionel deps from os-autoinst, and also all packages that are deps of other packages, so I think it's not possible at this step to fully generate it with my script.

I added an experimental `--check-ci` switch to the update-deps script, but I suggest we first try to manage the os-autonst deps with such a script too, and then have a further look on how we can manage the CI deps.

File: openQA.spec [hunk 5]
```diff
@@ -45,24 +45,32 @@
 %else
 %define python_scripts_requires %{nil}
 %endif
-%define assetpack_requires perl(Mojolicious::Plugin::AssetPack) >= 1.36, perl(CSS::Minifier::XS) perl(JavaScript::Minifier::XS)
-%define common_requires perl(Archive::Extract) perl(Config::IniFiles) perl(Cpanel::JSON::XS) perl(Cwd) perl(Data::Dump) perl(Data::Dumper) perl(Digest::MD5) perl(Getopt::Long) perl(Minion) >= 10.04, perl(Mojolicious) >= 8.24, perl(Try::Tiny) perl(Regexp::Common), perl(Storable)
+# The following line is generated from dependencies.yaml
+%define assetpack_requires perl(CSS::Minifier::XS) >= 0.01 perl(JavaScript::Minifier::XS) >= 0.11 perl(Mojolicious::Plugin::AssetPack) >= 1.36
```
Comment (okurz) [inline openQA.spec:49]: Last time I checked the commas to separate versioned tokens were necessary. Do you know how this works now? I can see https://build.opensuse.org/package/show/devel:openQA:TestGithub:PR-2934/openQA is fine but still :)

File: t/44-make-update-deps.t [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+
+if (@status) {
+    ok(0, "No changed files after '$make'");
+    diag @status;
+}
+else {
+    ok(1, "No changed files after '$make'");
+}
+
+
```
Comment (okurz) [inline t/44-make-update-deps.t:39]: How about

```perl
my @status = grep { not m/^\?/ } qx{git status --porcelain};
ok @status, "No changed files after '$make'" or diag @status;
```

File: tools/cpanfile.additions [hunk 7]
```diff
@@ -0,0 +1,5 @@
+# Additional hardcoded lines for cpanfile
```
Comment (okurz) [inline tools/cpanfile.additions:1]: can you say what makes these lines "hardcoded"? Why does the yaml file not work for specifying dependencies?

File: t/44-make-update-deps.t (hunk 6 above)
Comment (kraih) [reply t/44-make-update-deps.t:39]: You want `scalar @status` there.

Comment (perlpunk) [reply t/44-make-update-deps.t:39]: Probably even `not scalar @status` ;-)

**Comment by perlpunk** [comment]
> As there are some changes in dependencies in openQA.spec as well as cpanfile I recommend we separate the according commits which you already have separated, e.g. the "order …" ones into separate PRs which we can merge independantly for easier testing in production and potential reverting of smaller diffs.

Ok, do you want three several PRs for all three of those commits, or one?

**Comment by okurz** [comment]
three separate PRs please

File: openQA.spec [hunk 8]
```diff
@@ -45,23 +45,31 @@
...
 
 # All requirements needed by the tests executed during build-time.
 # Do not require on this in individual sub-packages except for the devel
 # package.
-%define test_requires %common_requires %main_requires %python_scripts_requires %worker_requires ShellCheck curl jq os-autoinst-devel perl(App::cpanminus) perl(Perl::Critic) perl(Perl::Critic::Freenode) perl(Selenium::Remote::Driver) perl(Selenium::Remote::WDKeys) perl(Test::Fatal) perl(Test::MockModule) perl(Test::Mojo) perl(Test::More) perl(Test::Output) perl(Test::Pod) perl(Test::Strict) perl(Test::Warnings) postgresql-server python3-setuptools python3-yamllint
+# The following line is generated from dependencies.yaml
+%define test_requires %common_requires %main_requires %python_scripts_requires %worker_requires ShellCheck curl jq os-autoinst-devel perl(App::cpanminus) perl(Perl::Critic) perl(Perl::Critic::Freenode) perl(Perl::Tidy) == 20200110 perl(Selenium::Remote::Driver) >= 1.23 perl(Selenium::Remote::WDKeys) perl(Test::Fatal) perl(Test::MockModule) perl(Test::Mojo) perl(Test::More) perl(Test::Output) perl(Test::Pod) perl(Test::Strict) perl(Test::Warnings) postgresql-server python3-setuptools python3-yamllint
```
Comment (okurz) [inline openQA.spec:66]: I see a (minor) problem here. You add Perl::Tidy which should IMHO only be required in `devel_requires` as we exclude any tidy check in the OBS tests

File: t/44-make-update-deps.t (hunk 6 above)
Comment (perlpunk) [reply t/44-make-update-deps.t]: updated

File: openQA.spec (hunk 5 above)
Comment (perlpunk) [reply openQA.spec:49]: I tested this syntax with another package in OBS, and also increased the version to check if it would correctly return unresolvable, and it did. so it seems to work.

File: openQA.spec (hunk 8 above)
Comment (perlpunk) [reply openQA.spec:66]: ok, Perl::Tidy was defined in `cpanfile` under `test`. Guess it's time to add a `devel` target also in that file.

Comment (perlpunk) [reply openQA.spec]: done

**Comment by perlpunk** [comment]
I also updated `.circleci/build_dependencies.sh` to work on `dependencies.yaml`.

**Comment by perlpunk** [comment]
> is the dependencies text file for circle CI obsolete then?

Unfortunately not (yet). 
See my comment https://github.com/os-autoinst/openQA/pull/2934#issuecomment-615211670

My update to `.circleci/build_dependencies.sh` was just regarding the Perl::Tidy version.

File: .circleci/build_dependencies.sh [hunk 9]
```diff
@@ -47,6 +47,7 @@ comm -13 gendep_before.txt gendep_after.txt | grep -v gpg-pubkey | grep -v openQ
 # let's tidy if Tidy version changes
 newtidyver="$(git diff $thisdir/dependencies.txt | grep perl-Perl-Tidy | grep '^+' | grep -o '[0-9]*' || :)"
 [ -z "$newtidyver" ] || {
-    sed -i -e "s/\('Perl::Tidy',\s\+'==\s\)\([0-9]\+\)\(.*\)/\1$newtidyver\3/g" cpanfile
+    sed -i -e "s/\(Perl::Tidy):\s\+'==\s\)\([0-9]\+\)\(.*\)/\1$newtidyver\3/g" dependencies.yaml
+    docker exec -t gendep make update-deps
```
Comment (kalikiana) [inline .circleci/build_dependencies.sh:51]: Why is this needed now? :thinking:

Comment (perlpunk) [reply .circleci/build_dependencies.sh:51]: Because `make update-deps` will update the spec and cpanfile.

Comment (kalikiana) [reply .circleci/build_dependencies.sh:51]: I was missing the forest for the tree I guess :-P Yeah, that makes sense of course. Previously it was just patching the cpanfile directly
